File: packages/go/accounts.go
package rdk

import (
	"crypto/sha256"
	"errors"
	"fmt"
	"strings"

	"github.com/cosmos/go-bip39"
	"github.com/decred/dcrd/dcrec/secp256k1/v4"
	"github.com/tyler-smith/go-bip32"
	"golang.org/x/crypto/ripemd160" //nolint:staticcheck // ripemd160 is required for Cosmos address derivation
)

// Account is a derived QoreChain native account. The private key material is
// returned explicitly and is never logged.
type Account struct {
	// Address is the bech32 (qor) account address.
	Address string
	// PublicKey is the 33-byte compressed secp256k1 public key.
	PublicKey []byte
	// PrivateKey is the 32-byte secp256k1 private key.
	PrivateKey []byte
}

// GenerateMnemonic generates a fresh BIP-39 English mnemonic. strengthBits of
// 128 yields 12 words (default), 256 yields 24 words.
func GenerateMnemonic(strengthBits int) (string, error) {
	if strengthBits == 0 {
		strengthBits = 128
	}
	entropy, err := bip39.NewEntropy(strengthBits)
	if err != nil {
		return "", err
	}
	return bip39.NewMnemonic(entropy)
}

// ValidateMnemonic validates a BIP-39 mnemonic against its checksum. It never
// errors; it returns false for invalid input.
func ValidateMnemonic(mnemonic string) bool {
	return bip39.IsMnemonicValid(mnemonic)
}

// hardened returns the hardened child index for BIP-32 derivation.
func hardened(i uint32) uint32 { return bip32.FirstHardenedChild + i }

// DeriveNativeAccount derives a native QoreChain account (Cosmos-style
// secp256k1) from a mnemonic at BIP-44 path m/44'/118'/0'/0/index. The address
// is the bech32 (qor) encoding of ripemd160(sha256(compressedPublicKey)).
func DeriveNativeAccount(mnemonic string, index uint32) (Account, error) {
	if !ValidateMnemonic(mnemonic) {
		return Account{}, errors.New("invalid mnemonic")
	}
	seed := bip39.NewSeed(mnemonic, "")
	master, err := bip32.NewMasterKey(seed)
	if err != nil {
		return Account{}, err
	}
	// m/44'/118'/0'/0/index
	path := []uint32{hardened(44), hardened(118), hardened(0), 0, index}
	key := master
	for _, p := range path {
		key, err = key.NewChildKey(p)
		if err != nil {
			return Account{}, fmt.Errorf("failed to derive secp256k1 key: %w", err)
		}
	}
	priv := key.Key // 32-byte private key
	privKey := secp256k1.PrivKeyFromBytes(priv)
	compressed := privKey.PubKey().SerializeCompressed() // 33 bytes

	sha := sha256.Sum256(compressed)
	rip := ripemd160.New()
	rip.Write(sha[:])
	digest := rip.Sum(nil) // 20 bytes

	address, err := bech32Encode(AccountPrefix, digest)
	if err != nil {
		return Account{}, err
	}
	return Account{
		Address:    address,
		PublicKey:  compressed,
		PrivateKey: append([]byte(nil), priv...),
	}, nil
}

// SignerFromEnv builds an operator Account from the environment, preferring a
// hex private key (QORE_OPERATOR_PRIVATE_KEY_HEX) over a mnemonic
// (QORE_MNEMONIC). The prefix selects the bech32 address prefix (empty defaults
// to the account prefix).
//
// It returns (Account{}, nil, false) when neither variable is set, so callers
// can give a friendly message; it returns a non-nil error when a variable is set
// but malformed.
func SignerFromEnv(env map[string]string, prefix string) (Account, error, bool) {
	hex := strings.TrimSpace(env["QORE_OPERATOR_PRIVATE_KEY_HEX"])
	mnemonic := strings.TrimSpace(env["QORE_MNEMONIC"])
	if prefix == "" {
		prefix = AccountPrefix
	}
	if hex != "" {
		priv, err := HexToBytes(hex)
		if err != nil {
			return Account{}, err, true
		}
		acc, err := AccountFromPrivateKey(priv, prefix)
		if err != nil {
			return Account{}, err, true
		}
		return acc, nil, true
	}
	if mnemonic != "" {
		acc, err := DeriveNativeAccount(mnemonic, 0)
		if err != nil {
			return Account{}, err, true
		}
		if prefix != AccountPrefix {
			acc, err = AccountFromPrivateKey(acc.PrivateKey, prefix)
			if err != nil {
				return Account{}, err, true
			}
		}
		return acc, nil, true
	}
	return Account{}, nil, false
}

// AccountFromPrivateKey builds an Account from a 32-byte secp256k1 private key
// and the given bech32 prefix (empty means the account prefix).
func AccountFromPrivateKey(priv []byte, prefix string) (Account, error) {
	if len(priv) != 32 {
		return Account{}, fmt.Errorf("private key must be 32 bytes, got %d", len(priv))
	}
	if prefix == "" {
		prefix = AccountPrefix
	}
	privKey := secp256k1.PrivKeyFromBytes(priv)
	compressed := privKey.PubKey().SerializeCompressed()
	sha := sha256.Sum256(compressed)
	rip := ripemd160.New()
	rip.Write(sha[:])
	digest := rip.Sum(nil)
	address, err := bech32Encode(prefix, digest)
	if err != nil {
		return Account{}, err
	}
	return Account{Address: address, PublicKey: compressed, PrivateKey: append([]byte(nil), priv...)}, nil
}

File: packages/jvm/src/main/java/io/github/qorechain/rdk/accounts/Accounts.java
package io.github.qorechain.rdk.accounts;

import io.github.qorechain.rdk.config.Constants;
import io.github.qorechain.rdk.util.Bech32;
import io.github.qorechain.rdk.util.Bytes;
import java.math.BigInteger;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import org.bouncycastle.crypto.digests.RIPEMD160Digest;
import org.web3j.crypto.Bip32ECKeyPair;
import org.web3j.crypto.MnemonicUtils;
import org.web3j.crypto.Sign;

/**
 * Account and signing helpers.
 *
 * <p>A native QoreChain account is derived from a BIP-39 mnemonic: BIP-39 seed → BIP-44 path
 * {@code m/44'/118'/0'/0/index} → compressed secp256k1 public key → {@code ripemd160(sha256(pubkey))}
 * → bech32 ({@code qor}) address.
 */
public final class Accounts {
    private Accounts() {}

    /**
     * Validate a BIP-39 mnemonic against its checksum. Never throws; returns false for invalid
     * input.
     */
    public static boolean validateMnemonic(String mnemonic) {
        return MnemonicUtils.validateMnemonic(mnemonic);
    }

    private static byte[] sha256(byte[] data) {
        try {
            return MessageDigest.getInstance("SHA-256").digest(data);
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException("SHA-256 unavailable", e);
        }
    }

    private static byte[] ripemd160(byte[] data) {
        RIPEMD160Digest d = new RIPEMD160Digest();
        d.update(data, 0, data.length);
        byte[] out = new byte[d.getDigestSize()];
        d.doFinal(out, 0);
        return out;
    }

    /** Left-pad/truncate a BigInteger to a fixed-length big-endian byte array. */
    static byte[] toFixed(BigInteger value, int length) {
        byte[] raw = value.toByteArray();
        byte[] out = new byte[length];
        if (raw.length == length) {
            return raw;
        }
        if (raw.length == length + 1 && raw[0] == 0) {
            System.arraycopy(raw, 1, out, 0, length);
            return out;
        }
        if (raw.length < length) {
            System.arraycopy(raw, 0, out, length - raw.length, raw.length);
            return out;
        }
        // raw longer than length: take the least-significant `length` bytes.
        System.arraycopy(raw, raw.length - length, out, 0, length);
        return out;
    }

    /** The 33-byte compressed secp256k1 public key for a private key. */
    static byte[] compressedPublicKey(BigInteger privateKey) {
        return Sign.publicPointFromPrivate(privateKey).getEncoded(true);
    }

    private static Account build(BigInteger privateKey, String prefix) {
        byte[] priv = toFixed(privateKey, 32);
        byte[] compressed = compressedPublicKey(privateKey);
        byte[] digest = ripemd160(sha256(compressed));
        String address = Bech32.encode(prefix, digest);
        return new Account(address, compressed, priv);
    }

    /**
     * Derive a native QoreChain account (Cosmos-style secp256k1) from a mnemonic at BIP-44 path
     * {@code m/44'/118'/0'/0/index}.
     */
    public static Account deriveNativeAccount(String mnemonic, int index) {
        if (!validateMnemonic(mnemonic)) {
            throw new IllegalArgumentException("invalid mnemonic");
        }
        byte[] seed = MnemonicUtils.generateSeed(mnemonic, "");
        Bip32ECKeyPair master = Bip32ECKeyPair.generateKeyPair(seed);
        int[] path = {
            44 | Bip32ECKeyPair.HARDENED_BIT,
            118 | Bip32ECKeyPair.HARDENED_BIT,
            0 | Bip32ECKeyPair.HARDENED_BIT,
            0,
            index
        };
        Bip32ECKeyPair derived = Bip32ECKeyPair.deriveKeyPair(master, path);
        return build(derived.getPrivateKey(), Constants.ACCOUNT_PREFIX);
    }

    /** Derive the default operator account ({@code index = 0}). */
    public static Account deriveNativeAccount(String mnemonic) {
        return deriveNativeAccount(mnemonic, 0);
    }

    /**
     * Build an {@link Account} from a 32-byte secp256k1 private key and the given bech32 prefix
     * (empty → account prefix).
     */
    public static Account accountFromPrivateKey(byte[] priv, String prefix) {
        if (priv.length != 32) {
            throw new IllegalArgumentException("private key must be 32 bytes, got " + priv.length);
        }
        String p = (prefix == null || prefix.isEmpty()) ? Constants.ACCOUNT_PREFIX : prefix;
        return build(new BigInteger(1, priv), p);
    }

    /** The outcome of resolving a signer from the environment. */
    public static final class SignerResult {
        /** The derived account, or {@code null} when no signer is configured. */
        public final Account account;
        /** True when a variable was present (even if it was malformed). */
        public final boolean present;

        SignerResult(Account account, boolean present) {
            this.account = account;
            this.present = present;
        }
    }

    /**
     * Build an operator {@link Account} from the environment, preferring a hex private key
     * ({@code QORE_OPERATOR_PRIVATE_KEY_HEX}) over a mnemonic ({@code QORE_MNEMONIC}). The prefix
     * selects the bech32 address prefix (empty → account prefix).
     *
     * <p>Returns a result with {@code present = false} when neither variable is set, so callers can
     * give a friendly message; throws when a variable is set but malformed.
     */
    public static SignerResult signerFromEnv(java.util.Map<String, String> env, String prefix) {
        String hex = trim(env.get("QORE_OPERATOR_PRIVATE_KEY_HEX"));
        String mnemonic = trim(env.get("QORE_MNEMONIC"));
        String p = (prefix == null || prefix.isEmpty()) ? Constants.ACCOUNT_PREFIX : prefix;
        if (!hex.isEmpty()) {
            byte[] priv = Bytes.hexToBytes(hex);
            return new SignerResult(accountFromPrivateKey(priv, p), true);
        }
        if (!mnemonic.isEmpty()) {
            Account acc = deriveNativeAccount(mnemonic, 0);
            if (!p.equals(Constants.ACCOUNT_PREFIX)) {
                acc = accountFromPrivateKey(acc.privateKey, p);
            }
            return new SignerResult(acc, true);
        }
        return new SignerResult(null, false);
    }

    private static String trim(String s) {
        return s == null ? "" : s.trim();
    }
}
